### ISSUE TITLE
A float32 forward earlier in the process shifts a later float64 forward by 1.4e-9: a precision-blind cache on the forward path (#1206)

### ISSUE THREAD

**Issue by suchethac** [suchethac/tengri#2275]
**Child of #1206. Same class as #1392 (structural-kernel cache not keyed on precision) and #2024 (precomp cache not keyed on dtype), on the forward this time.**

A float32 forward of a model earlier in a process changes a later **float64** forward of the same model by ~1.4e-9. Measured on `origin/main` at 18cf9fb9e (CPU, jaxlib 0.11.0), kubota_done + SKIRTOR composable AGN, `agn_log_mbh=6`, `agn_log_lbol=11` (the #2210 test model):

| process | float64 `rest_sed.sum()` | hex |
|---|---|---|
| float64 only | `1.5567108723215032e+32` | `0x1.eb361a52c8792p+106` |
| float32 forward first, then float64 | `1.5567108701948514e+32` | `0x1.eb361a4786529p+106` |

relative difference **1.366e-9** (at `agn_log_mbh` 8 and 10: 1.58e-10 and 7.98e-10). Both float64 arms use `with jax.enable_x64(True)` and `jnp.float64` inputs; only the history of the process differs. Photometry through `sdss_r`/`wise_w3`/`wise_w4` moves too.

## Why it matters

- Every `tests/regression/precision` file runs float32 and float64 arms in one process; a float64 "reference" taken after a float32 arm is contaminated at the 1e-9 level, which is above the 1e-12 bar the tree uses for "float64 unchanged". PR #2270 first reported its regrouping as "moving float64 by 1.37e-9" for exactly this reason; measured in separate processes the regrouping is bit-identical. Any future regrouping PR will trip over the same thing.
- A user who fits in float32 for speed and then re-evaluates the best fit in float64 in the same session gets a float64 number that is not the float64 number.

## Not the cause

- `components/agn/_nthcomp.py` templates are loaded as float32 by construction and are the same in both arms.
- The persistent XLA compilation cache is keyed on the HLO (dtype included).
- `~/.cache/tengri_precomp` is keyed on dtype since #2024; the probe below reproduces unchanged with `TENGRI_DISABLE_PRECOMP_CACHE=1` (measured).

Candidates to bisect (in order of suspicion): a `functools.cache`/`lru_cache`d builder that captures the first-call precision (SKIRTOR grid interpolation tables, filter LUTs, `kd_precompute` tables, the dust age-weight sigmoid at `Model.__init__`), a module-level mutable that a float32 build overwrites, or JAX's own tracing cache for a `custom_jvp`/`custom_vjp` rule keyed without the x64 flag.

## Reproduce

```bash
JAX_PLATFORMS=cpu python xprec_probe.py f64only      # 1.5567108723215032e+32
JAX_PLATFORMS=cpu python xprec_probe.py f32then64    # 1.5567108701948514e+32
```
with `xprec_probe.py` building the model above and calling `model.predict` once under `jax.enable_x64(False)` with float32 inputs, then once under `jax.enable_x64(True)` with float64 inputs (script attached in the first comment). The effect is process-wide, not model-specific: in `tests/regression/precision/test_float32_scale_seam_sweep.py`, float32 tests of *other* models earlier in the file are enough to trigger it.

## Proposed fix

Bisect with the probe by clearing candidate caches between the two arms (`jax.clear_caches()` between the two arms does not restore the clean value: measured, same hex); once the cache is found, key it on `jax.config.jax_enable_x64` / the array dtype like #1392 did for the kernel cache. Guard: a test that runs float32 then float64 in one process and asserts the float64 result equals a fresh-process float64 result to `rtol=1e-12` (the subprocess pattern now used by `test_agn_black_hole_mass_float64_is_unchanged_by_the_2210_regrouping` in #2270).

**Comment by suchethac** [comment]
Probe script used for the measurements above (`xprec_probe.py`):

```python
"""Does a float32 forward in the same process change a later float64 forward?

Usage: python xprec_probe.py {f64only|f32then64}
Prints repr of the float64 rest_sed sum at agn_log_mbh=6 for the #2270 test model.
"""

import sys

import jax
import jax.numpy as jnp
import numpy as np

import tengri
from tengri import DEFAULT, Fixed, Observation, Photometry, SEDModel, Uniform
from tengri.components.stellar.sps.dsps_wrapper import load_ssp_data

MODE = sys.argv[1]
print("tengri from", tengri.__file__)
ssp = load_ssp_data("data/fsps_prsc_miles_chabrier.h5")


def model():
    return SEDModel.build(
        ssp_data=ssp,
        observation=Observation(photometry=Photometry.from_names(["sdss_r", "wise_w3", "wise_w4"])),
        sfh={"type": "delayed", "all_params": Fixed(DEFAULT), "log_total_mass": Uniform(9.0, 11.0),
             "tau_gyr": 1.0, "age_gyr": 5.0},
        dust_attenuation={"type": "two_component", "law": "calzetti", "all_params": Fixed(DEFAULT),
                          "tau_diff": 0.3, "tau_bc": 0.0},
        agn={"type": "composable", "all_params": Fixed(DEFAULT),
             "disc": {"type": "kubota_done", "all_params": Fixed(DEFAULT)},
             "torus": {"type": "skirtor", "all_params": Fixed(DEFAULT)},
             "norm": "cigale_joint", "log_lbol": Fixed(11.0), "fracAGN": 0.1},
        redshift=Fixed(0.1),
    )


def predict(dtype):
    with jax.enable_x64(dtype is jnp.float64):
        m = model()
        p = {"sfh_delayed_log_total_mass": jnp.asarray(10.0, dtype=dtype),
             "agn_log_mbh": jnp.asarray(6.0, dtype=dtype)}
        pred = m.predict(p)
        return np.asarray(pred.rest_sed()), np.asarray(pred.photometry())


if MODE.startswith("f32then64"):
    r32, _ = predict(jnp.float32)
    print("f32 sum", repr(float(r32.sum())))
    if MODE == "f32then64_clear":
        jax.clear_caches()
        print("cleared jax caches")
r64, ph = predict(jnp.float64)
print("f64 sum", repr(float(r64.sum())), "hex", float(r64.sum()).hex())
print("f64 phot", [repr(float(x)) for x in ph])
```

**Comment by suchethac** [comment]
Independent reproduction on another machine (M4 Pro, CPU backend, jax 0.11.1, `TENGRI_DISABLE_JAX_CACHE=1 TENGRI_DISABLE_PRECOMP_CACHE=1`, `PYTHONPATH` pinned per tree), same probe:

| tree | f64 only | f32 then f64 | rel |
|---|---|---|---|
| origin/main f6c36d9d3 | `1.5567108722981187e+32` (`0x1.eb361a52a8c80p+106`) | `1.5567108701714671e+32` (`0x1.eb361a4766a18p+106`) | 1.366e-9 |
| #2271 branch 46fcae1c3 (all module-scope constants host-side) | `1.556710872298117e+32` | `1.5567108701714651e+32` | 1.366e-9 |

So the shift is independent of jaxlib version (0.11.0 here, 0.11.1 there) and of the module-scope-constant class (#2271 does not change it): it is a tengri-level object cached without a dtype key. (The ~1.5e-11 difference between the two machines' float64-only values is the usual cross-machine kernel ordering, unrelated.)
